synapse/storage/databases/main/roommember.py

```diff
@@ -1057,21 +1057,25 @@ async def get_current_hosts_in_room(self, room_id: str) -> Set[str]:
         users = self.get_users_in_room.cache.get_immediate(
             (room_id,), None, update_metrics=False
         )
-        if users is not None:
-            # Because `users` is sorted from lowest -> highest depth, the set of
-            # domains will also be sorted that way.
-            return {get_domain_from_id(u) for u in users}
-
-        if isinstance(self.database_engine, Sqlite3Engine):
+        if users is None and isinstance(self.database_engine, Sqlite3Engine):
             # If we're using SQLite then let's just always use
             # `get_users_in_room` rather than funky SQL.
             users = await self.get_users_in_room(room_id)
-            return {get_domain_from_id(u) for u in users}
+
+        if users is not None:
+            # Because `users` is sorted from lowest -> highest depth, the list
+            # of domains will also be sorted that way.
+            domains: List[str] = []
+            for u in users:
+                domain = get_domain_from_id(u)
+                if domain not in domains:
+                    domains.append(domain)
```

Comment:
This will be O(n^2) slow because `domain not in domains` is an expensive `O(n)` thing to do.
A technique you might want to use is to maintain a set of the domains you've already seen...

```suggestion
            domains: List[str] = []
            processed_domains: Set[str] = set()
            for u in users:
                domain = get_domain_from_id(u)
                if domain not in processed_domains:
                    domains.append(domain)
                    processed_domains.add(domain)
```

Reply:
Any preference on having a separate `Set` for fast lookups vs using a `Dict` which you mentioned maintains insertion order and just call `dict.values()` at the end?

Reply:
I take it you mean `.keys()`, but I think `Dict` maintaining insertion order is a CPython quirk and not guaranteed by the Python standard, so I'd rather not rely on it. We do have some admins running unusual interpreters :-)

Reply:
> `Dict` maintaining insertion order is a CPython quirk and not guaranteed by the Python standard

It's guaranteed from 3.7 upwards. See "Python data model improvements" in https://docs.python.org/3/whatsnew/3.7.html#summary-release-highlights

Reply:
> It's guaranteed from 3.7 upwards (which is the oldest Python we support). See "Python data model improvements" in [docs.python.org/3/whatsnew/3.7.html#summary-release-highlights](https://docs.python.org/3/whatsnew/3.7.html#summary-release-highlights)

good to know. In that case, feel free to do whichever is best since 3.7 is our minimum version.